js/ui/shellDBus.js

```diff
@@ -542,6 +546,21 @@ const AppStoreService = new Lang.Class({
             IconGridLayout.layout.appendIcon(id, IconGridLayout.DESKTOP_GRID_ID);
     },
 
+    ReplaceAppIfPossible: function(originalId, replacementId) {
+        let eventRecorder = EosMetrics.EventRecorder.get_default();
+        let appId = new GLib.Variant('s', replacementId);
+        eventRecorder.record_event(SHELL_APP_ADDED_EVENT, appId);
+
+        // Can't replace a folder
+        if (IconGridLayout.layout.iconIsFolder(originalId))
+            return;
+
+        // We can just replace the app icon directly now,
+        // since the replace operation degenerates to
+        // append if the source icon was not available
+        IconGridLayout.layout.replaceIcon(originalId, replacementId, IconGridLayout.DESKTOP_GRID_ID);
```

Comment:
That is a good point. Do you know if there is a way to look up the folder an app is already in?